internal/ansible/ansible.go

```diff
@@ -0,0 +1,196 @@
+	EventData   struct {
+		CRCDispatcherCorrelationID string `json:"crc_dispatcher_correlation_id"`
+		CRCDispatcherErrorCode     string `json:"crc_dispatcher_error_code"`
+		CRCDispatcherErrorDetails  string `json:"crc_dispatcher_error_details"`
+	} `json:"event_data"`
+}
+
+// RunPlaybook creates an ansible-runner job to run the provided playbook. The
+// function returns a channel over which the caller can receive ansible-runner
+// events as they happen. The channel is closed when the job completes.
+func RunPlaybook(id string, playbook []byte, correlationID string) (chan Event, error) {
+	privateDataDir := filepath.Join(constants.StateDir, id)
+
+	if err := os.WriteFile(filepath.Join(constants.StateDir, id+".yaml"), playbook, 0600); err != nil {
+		return nil, fmt.Errorf("cannot write playbook to temp file: %v", err)
+	}
+	if err := os.MkdirAll(filepath.Join(privateDataDir, "artifacts", id, "job_events"), 0755); err != nil {
+		return nil, fmt.Errorf("cannot create private state dir: %v", err)
+	}
+
+	events := make(chan Event)
+
+	go func() {
+		// publish an "executor_on_start" event to signal cloud connector that a run
+		// event has started.
+		events <- Event{
+			Event:     "executor_on_start",
+			UUID:      uuid.New().String(),
+			Counter:   -1,
+			Stdout:    "",
+			StartLine: 0,
+			EndLine:   0,
+			EventData: struct {
+				CRCDispatcherCorrelationID string `json:"crc_dispatcher_correlation_id"`
+				CRCDispatcherErrorCode     string `json:"crc_dispatcher_error_code"`
+				CRCDispatcherErrorDetails  string `json:"crc_dispatcher_error_details"`
+			}{
+				CRCDispatcherCorrelationID: correlationID,
+			},
+		}
+	}()
+
+	err := exec.StartProcess(
+		"/usr/bin/python3",
+		[]string{
+			"-m",
+			"ansible_runner",
+			"run",
+			"--ident",
+			id,
+			"--json",
+			"--playbook",
+			filepath.Join(constants.StateDir, id+".yaml"),
+			privateDataDir,
+		},
+		[]string{
+			"PATH=/sbin:/bin:/usr/sbin:/usr/bin",
+			"PYTHONPATH=" + filepath.Join(constants.LibDir, "rhc-worker-playbook"),
+			"ANSIBLE_COLLECTIONS_PATH=" + filepath.Join(
+				constants.DataDir,
+				"rhc-worker-playbook",
+				"ansible",
+				"collections",
+				"ansible_collections",
+			),
+		},
+		func(pid int, stdout, stderr io.ReadCloser) {
+			slog.Debug("process started", "pid", pid, "runner_ident", id)
+
+			// start a goroutine to watch for event files being written to the
+			// job_events directory.
+			go func() {
+				c := make(chan notify.EventInfo, 1)
+				if err := notify.Watch(filepath.Join(privateDataDir, "artifacts", id, "job_events"), c, notify.InMovedTo); err != nil {
+					slog.Error(
+						"failed to watch private data dir",
+						"dir",
+						privateDataDir,
+						"err",
+						err,
+					)
+					return
+				}
+				defer notify.Stop(c)
+
+				for e := range c {
+					slog.Debug("notify event", "event", e.Event(), "path", e.Path())
+					switch e.Event() {
+					case notify.InMovedTo:
+						if strings.HasSuffix(e.Path(), ".json") {
```

Comment:
I don't feel like it's necessary to debug-print this, since we're already
debug-printing each event. Maybe a happy-path info message like "sent event"
after the event gets sent to the channel?